### PR TITLE
[3.3.5] [WIP] [HELP NEEDED] Scripts/Quest: Look at the Size of It!

### PR DESCRIPTION
**Pull request by PolarCookie**
**Changes proposed:**
Scripted the event that happens when you loot a Dark Iron Pillow in the Slag Pit, Searing Gorge.
2 waves of 10 Sleepy Dark Iron Dwarves spawn, then Chambermaid Pillaclencher whom drops the Chambermaid Pillaclencher's Pillow.

Disclaimer: I never did this quest back on retail, used the following relevant links:
http://www.wowhead.com/quest=7702/kill-em-with-sleep-deprivation
http://www.wowhead.com/quest=7704/look-at-the-size-of-it
http://www.wowhead.com/quest=28057/kill-em-with-sleep-deprivation

http://www.wowhead.com/npc=14635/sleepy-dark-iron-worker
http://www.wowhead.com/npc=14636/chambermaid-pillaclencher

http://www.wowhead.com/object=179832/pillaclenchers-ornate-pillow

https://www.youtube.com/watch?v=Gu29CrLxFN0
https://youtu.be/RcD1Ws92u5U?t=35m
https://www.youtube.com/watch?v=S1lPtS-GkwA

Help and feedback appreciated.

**Target branch(es):**
3.3.5

**Tests performed:**
Builds and works for me.

**Known issues and TODO list:**
- [ ] Find some sniffed coordinates.
- [ ] Research Sleepy Dark Iron Dwarves (14635) and Chambermaid Pillaclencher's (14636) melee damage, Chambermaid is supposed to hit for 20-40, right now 500-600 on cloth.
- [ ] Research the cooldowns on spells Smother (23417) and Pillow Fight (23416).
- [ ] Research Pillow (179828) placement and respawn time; 8 pillows (179828) on a 5 min timer currently.
- [x] Sleepy Dark Iron Dwarves (14635) should run up and down the corridor.

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    uint32_t duration;
+    uint32_t numWave;
+    bool running;
+
+    std::mutex triggerLock;
+};
+
+static event_dark_iron_rush *edir_p = NULL;
```
Comment (ccrs) [inline src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp]: probably there is a spell that launches the event, maybe its server-side

Comment (PolarCookie) [reply src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp]: https://wotlk.evowow.com/?object=179829
???

Comment (illfated) [reply src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp]: (for those of you using ad block software : )
https://woehead.way-of-elendil.fr/?object=179829
(Dark Iron Pillow Event Generator)

File: src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    go_dark_iron_pillow() : GameObjectScript("go_dark_iron_pillow") {}
+
+
+    struct go_dark_iron_pillowAI : public GameObjectAI
+    {
+        go_dark_iron_pillowAI(GameObject* go) : GameObjectAI(go) {};
+
+        bool GossipHello(Player* player) override
+        {
+            std::lock_guard<std::mutex> triggerGuard(triggerLock);
```
Comment (PolarCookie) [inline src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp]: Do I need this lock?
What's the threading model for scripts, what happens if 2 players click a different pillow simultaneously?

**Comment by Killyana** [comment]
Can you provide on TODO the IDs of npcs, gobs, spells ect..

File: src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+* Free Software Foundation; either version 2 of the License, or (at your
+* option) any later version.
+*
+* This program is distributed in the hope that it will be useful, but WITHOUT
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+* more details.
+*
+* You should have received a copy of the GNU General Public License along
+* with this program. If not, see <http://www.gnu.org/licenses/>.
```
Comment (illfated) [inline src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp]: Cosmetic, but required codestyle: 1 leading space indent on lines 2-16 (vertical alignment)

**Comment by illfated** [comment]
I think Killyana meant something like this:

- [ ] research Sleepy Dark Iron Dwarves' (14635) and Chambermaid Pillaclencher's melee damage, Chambermaid (14636) is supposed to hit for 20-40, right now 500-600 on cloth.
- [ ] research the cooldowns on spells Smother (23417) and Pillow Fight (23416).
- [ ] research Pillow (179828) placement and respawn time; 8 pillows (179828) on a 5 min timer currently.
- [X] Sleepy Dark Iron Dwarves (14635) should run up and down the corridor.

File: src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                        {
+                            summon->ToCreature()->GetMotionMaster()->MovePath(NPC_SLEEPY_DARK_IRON_WORKER * 100+10+i, true);
+
+                            if (i == 0)
+                                summon->ToCreature()->AI()->Talk(1);
+                        }
+                    }
+                    break;
+                case EVENT_CHAMBERMAID:
+                    TempSummon* summon = owner->SummonCreature(NPC_CHAMBERMAID_PILLACLENCHER, -6545.88f, -1345.29f, 208.89f,0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 300000);
```
Comment (illfated) [inline src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp]: unused variable, needs to be commented out:
```cpp
                    TempSummon* /*summon*/ = owner->SummonCreature(NPC_CHAMBERMAID_PILLACLENCHER, -6545.88f, -1345.29f, 208.89f,0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 300000);
```
(unless you have planned to add more code using `summon` below)

File: src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp (hunk 2 above)
Comment (illfated) [reply src/server/scripts/EasternKingdoms/zone_searing_gorge.cpp]: I don't know, but maybe some of our fellow users could test this for you.
I would have tested that if it wasn't for the fact that I play alone on my PC.
To me, this PR looks like long-missed good fun in the code. :)

**Comment by PolarCookie** [comment]
New Years house cleaning; I'm not getting back to this PR anytime soon, keep getting distracted, someone else can feel free to continue this.

**Comment by illfated** [comment]
Label: **[Closed-Abandoned]**

**Comment by offl** [comment]
https://github.com/TrinityCore/TrinityCore/commit/513cbd289a9b6ac6c120c5f93a5b18f4cc04693e

**Pull request closed without merging**